third_party/e2e/engine.cc

```diff
@@ -0,0 +1,90 @@
+#include "engine.h"
+
+#include <cstdint>
+
+#include "lib/translator/translation.h"
+#include "nvme_driver.h"
+
+namespace {
+
+uint64_t AllocPagesCallback(uint16_t count) { return AllocPages(count); }
+
+void DeallocPagesCallback(uint64_t addr, uint16_t count) {
+  DeallocPages(addr, count);
+}
+
+}  // namespace
+
+void SetEngineCallbacks(void) {
+  translator::SetDebugCallback(Print);
+  translator::SetAllocPageCallbacks(AllocPagesCallback, DeallocPagesCallback);
+}
+
+ScsiToNvmeResponse ScsiToNvme(unsigned char* cmd_buf, unsigned short cmd_len,
+                              unsigned long long lun, unsigned char* sense_buf,
+                              unsigned short sense_len, unsigned char* data_buf,
+                              unsigned short data_len, bool is_data_in) {
+  // Create translation object
+  translator::Translation translation;
+
+  // Package parameters and run translation begin
+  translator::Span<uint8_t> scsi_cmd(cmd_buf, cmd_len);
+  translator::Span<uint8_t> buffer(data_buf, data_len);
+  translator::BeginResponse begin_resp =
+      translation.Begin(scsi_cmd, buffer, lun);
+
+  if (begin_resp.status == translator::ApiStatus::kFailure) {
+    Print("Incorrect usage of Translation Library API");
```

Comment:
Is this guaranteed to be an incorrect usage from part of the caller? If not, the error message might be deceiving.